File: frontend/src/server/auth/permissions.ts
/**
 * Permission definitions and role-based mappings.
 * Defines what actions each role can perform.
 */

import { UserRole, Permission } from './types';

/**
 * Map of roles to their allowed permissions.
 * System admins have full access including namespace management.
 * Namespace admins have full access within their namespace.
 * Instructors have teaching access, students have limited access.
 */
export const ROLE_PERMISSIONS: Record<UserRole, Permission[]> = {
  'system-admin': [
    // Session permissions
    'session.create',
    'session.join',
    'session.viewAll',
    'session.viewOwn',
    'session.delete',

    // Class permissions
    'class.read',
    'class.create',
    'class.update',
    'class.delete',

    // Section permissions
    'section.read',
    'section.create',
    'section.update',
    'section.delete',

    // Problem permissions
    'problem.read',
    'problem.create',
    'problem.update',
    'problem.delete',

    // User management permissions
    'user.manage',
    'user.create',
    'user.delete',
    'user.viewAll',
    'user.changeRole',

    // Data access permissions
    'data.viewAll',
    'data.viewOwn',
    'data.export',

    // Namespace management permissions
    'namespace.create',
    'namespace.manage',
    'namespace.delete',
    'namespace.viewAll',

    // System administration
    'system.admin',
  ],

  'namespace-admin': [
    // Session permissions
    'session.create',
    'session.join',
    'session.viewAll',
    'session.viewOwn',
    'session.delete',

    // Class permissions
    'class.read',
    'class.create',
    'class.update',
    'class.delete',

    // Section permissions
    'section.read',
    'section.create',
    'section.update',
    'section.delete',

    // Problem permissions
    'problem.read',
    'problem.create',
    'problem.update',
    'problem.delete',

    // User management permissions
    'user.manage',
    'user.create',
    'user.delete',
    'user.viewAll',
    'user.changeRole',

    // Data access permissions
    'data.viewAll',
    'data.viewOwn',
    'data.export',
  ],

  instructor: [
    // Session permissions
    'session.create',
    'session.join',
    'session.viewAll',
    'session.viewOwn',
    'session.delete',

    // Class permissions
    'class.read',
    'class.create',
    'class.update',
    'class.delete',

    // Section permissions
    'section.read',
    'section.create',
    'section.update',
    'section.delete',

    // Problem permissions
    'problem.read',
    'problem.create',
    'problem.update',
    'problem.delete',

    // User management permissions
    'user.manage',
    'user.create',
    'user.delete',
    'user.viewAll',

    // Data access permissions
    'data.viewAll',
    'data.viewOwn',
    'data.export',
  ],

  student: [
    // Session permissions (limited)
    'session.join',
    'session.viewOwn',

    // Class/section permissions (read only - see sections they're enrolled in)
    'class.read',
    'section.read',

    // Problem permissions (read only - see problems assigned to them)
    'problem.read',

    // Data access permissions (own data only)
    'data.viewOwn',
  ],
};

/**
 * Check if a role has a specific permission.
 */
export function hasRolePermission(role: UserRole, permission: Permission | string): boolean {
  const rolePermissions = ROLE_PERMISSIONS[role];
  return rolePermissions.includes(permission as Permission);
}

/**
 * Get all permissions for a role.
 */
export function getRolePermissions(role: UserRole): Permission[] {
  return ROLE_PERMISSIONS[role];
}

/**
 * Permission descriptions for documentation/UI.
 */
export const PERMISSION_DESCRIPTIONS: Record<Permission, string> = {
  'session.create': 'Create new coding sessions',
  'session.join': 'Join existing coding sessions',
  'session.viewAll': 'View all coding sessions (including other instructors)',
  'session.viewOwn': 'View own coding sessions',
  'session.delete': 'Delete coding sessions',

  'class.read': 'View classes',
  'class.create': 'Create new classes',
  'class.update': 'Update class settings',
  'class.delete': 'Delete classes',

  'section.read': 'View sections',
  'section.create': 'Create new sections',
  'section.update': 'Update section settings',
  'section.delete': 'Delete sections',

  'problem.read': 'View problems',
  'problem.create': 'Create new problems',
  'problem.update': 'Update problems',
  'problem.delete': 'Delete problems',

  'user.manage': 'Manage user accounts',
  'user.create': 'Create new user accounts',
  'user.delete': 'Delete user accounts',
  'user.viewAll': 'View all user accounts',
  'user.changeRole': 'Change user roles',

  'data.viewAll': 'View all student data and code',
  'data.viewOwn': 'View own data and code',
  'data.export': 'Export data and analytics',

  'namespace.create': 'Create new namespaces',
  'namespace.manage': 'Manage namespace settings and users',
  'namespace.delete': 'Delete namespaces',
  'namespace.viewAll': 'View data across all namespaces',

  'system.admin': 'Full system administration access',
};

File: go-backend/internal/auth/permissions.go
package auth

// Permission represents a capability in the system.
// Permissions are assigned to roles via the rolePermissions map.
type Permission string

// Permissions (~10 capabilities vs coding-tool's 57).
// These are coarse-grained capabilities that match actual use cases.
const (
	// Session permissions
	PermSessionJoin   Permission = "session.join"   // Join a session as participant
	PermSessionManage Permission = "session.manage" // Create, update, delete sessions

	// Content permissions (consolidated CRUD for classes/sections/problems)
	PermContentManage Permission = "content.manage" // CRUD classes, sections, problems

	// Data access permissions
	PermDataViewOwn Permission = "data.viewOwn" // View own data (submissions, grades)
	PermDataViewAll Permission = "data.viewAll" // View all data in scope (instructor view)
	PermDataExport  Permission = "data.export"  // Export data (CSV, reports)

	// User management permissions
	PermUserManage     Permission = "user.manage"     // CRUD users within namespace
	PermUserChangeRole Permission = "user.changeRole" // Change user roles

	// Preview permissions
	PermPreviewStudent Permission = "preview.enter" // Enter preview-as-student mode

	// Administrative permissions
	PermNamespaceManage Permission = "namespace.manage" // Manage namespace settings
	PermSystemAdmin     Permission = "system.admin"     // Full system access
)

// rolePermissions maps each role to its granted permissions.
// Higher roles include all permissions of lower roles implicitly via this map.
var rolePermissions = map[Role][]Permission{
	RoleStudent: {
		PermSessionJoin,
		PermDataViewOwn,
	},
	RoleInstructor: {
		PermSessionJoin,
		PermSessionManage,
		PermContentManage,
		PermDataViewOwn,
		PermDataViewAll,
		PermDataExport,
		PermPreviewStudent,
	},
	RoleNamespaceAdmin: {
		PermSessionJoin,
		PermSessionManage,
		PermContentManage,
		PermDataViewOwn,
		PermDataViewAll,
		PermDataExport,
		PermUserManage,
		PermUserChangeRole,
		PermNamespaceManage,
		PermPreviewStudent,
	},
	RoleSystemAdmin: {
		PermSessionJoin,
		PermSessionManage,
		PermContentManage,
		PermDataViewOwn,
		PermDataViewAll,
		PermDataExport,
		PermUserManage,
		PermUserChangeRole,
		PermNamespaceManage,
		PermSystemAdmin,
		PermPreviewStudent,
	},
}

// permissionIndex is a pre-computed lookup for O(1) permission checks.
// Built at init time from rolePermissions.
var permissionIndex map[Role]map[Permission]struct{}

func init() {
	permissionIndex = make(map[Role]map[Permission]struct{}, len(rolePermissions))
	for role, perms := range rolePermissions {
		permissionIndex[role] = make(map[Permission]struct{}, len(perms))
		for _, p := range perms {
			permissionIndex[role][p] = struct{}{}
		}
	}
}

// HasPermission checks if a role has a specific permission.
func HasPermission(role Role, perm Permission) bool {
	perms, ok := permissionIndex[role]
	if !ok {
		return false
	}
	_, has := perms[perm]
	return has
}

// RolePermissions returns the permissions granted to a role.
// Returns nil for unknown roles.
func RolePermissions(role Role) []Permission {
	return rolePermissions[role]
}
